### DISCUSSION TITLE
Providing a better experience with JSON storage in property editors

### DISCUSSION THREAD

**Discussion by DanDiplo** [umbraco/Umbraco-CMS#10528] [Features and ideas]
More and more property editors are storing content as blobs of JSON. I've no argument against this, as it's a fast and very flexible way of storing complex models simply - it is easily referenced client-side in the back-office and can be de-serialised to strongly typed models for back-end code. Nested Content, Block List, Grid and even new Media Picker 3 now all store content as JSON.

However, there are a number of issues this causes and I'd like to open up a discussion in how best they can be solved. For me, the main issues are these:

### Lack of Referential Integrity

If, for instance, an Element Type is serialised to JSON within (say) the Block List editor there is currently no way of knowing which elements are stored where. There's no referential link between the element and the property editor.

To illustrate this, imagine I create a simple Element Type called **Person** with one property with an alias of `name` using a `Textstring`. I then use this as a block within  a Block List editor datatype. I then have a number of pages that I create where I add the **Person** block and populate the `Name` field content. If I subsequently go into the back office and edit the Element Type and change the `name` alias to `surname` I will find _all_ the content for every instance of the block will have gone in the backend. If I'd used this on every page in a site and it was a required field, it would mean I couldn't publish a single page until I repopulate all that content.

Actually, and this can be even worse, the content hasn't gone until the page is republished - it's still stored in the published content JSON as `name` - but it is no longer mapped back to the property alias `name` in the Umbraco UI. If in future I add another property with alias of `name` to the Element the content will be resurrected! But what happens if my new name property is using a datatype that doesn't expect a text string but, say, a UDI? I can get JSON errors which prevents me publishing the page at all.

Which leads to the second problem...

### JSON Conversion Errors

One of the nice things about Umbraco is you can swap the property editor used by a property at any time. This works great when the property editor stores the content as a string - you can swap a rich text editor for a textbox and the worst that can happen is your textbox will contain some HTML tags. However, when the property editor stores content as JSON this can cause issues if the property editor expects the content to be serialised in a specific format.

For instance, if you have a **Block List editor** and decide you want to swap it for **Nested Content** (which doesn't seem unreasonable, as you might expect them to work in a similar way) then not only do you lose all the content but when you attempt to publish the page you'll be greeted to a full-screen error dialogue similar to the one below:

![image](https://user-images.githubusercontent.com/6676168/123282605-93b9a480-d502-11eb-8d00-84793faa2c87.png)

This will completely prevent you from publishing the page. The only way to fix it is to remove the property or at least change the property editor to something that doesn't parse JSON.

I understand why Umbraco does it, but is this the most friendly experience? After all Umbraco doesn't warn you when changing the property editor and then leaves you in a state where your entire editing experience is completely broken. Is there a way that this could be handled more gracefully?

- Could there be a warning if you try and swap to an incompatible property editor?
- Could the above error dialogue just be a publish error notification with the full error logged?
- When swapping property editors could there be an option to "blank out" the existing value, so they are nulled, removing any existing JSON?
- Could the property editor be smart and try and convert from similar formats? For instance, if it can't be parsed as as  `BlockListModel` could it be parsed as a `Nested Content` model and converted on save? This would have been handy for the new Media Picker 3, for instance, which suffers from this same issue.

I don't know all of the answers, which is why I'm opening this for discussion, as I'm sure we have some smart people who can give this consideration.

As more and more complex property editors are developed then I think we need to find a way to solve these core problems which can at best be annoying and at worst break entire sites or end up with large swathes of content getting lost.

**Comment by skttl** [comment]
Could be interesting to do a concept of data editor mappers, where you could define which data editor you are migrating from, and which you are migrating to. When you attempt to change the data editor of a data type, the UI could warn you if no mappers exist, and give you the option to null existing values.

I think that would be very friendly.

Could also help you migrate from older editors in old projects, eg grid to block list. That would require you to add your own mapper though.

But core could come with mappers for media picker 2 to media picker 3, nested content to Block list, content picker to url picker etc.

**Reply by DanDiplo** [reply]
I think that's a great idea - love it!

**Reply by skttl** [reply]
Following up, when changing the dropdown value for selecting the property editor
![image](https://user-images.githubusercontent.com/3726467/123384246-746b5780-d594-11eb-826f-c4d11884fce2.png)

An overlay could pop up, warning the user of potential data corruption, or data loss when changing the property editor. It could then list some options, based on what Data Editor Mappers are in the project. I envision Umbraco to come with a selection of useful mappers, like Nested Content to Block List, MediaPicker 2 to MediaPicker 3 etc, but also a Reset mapper, and a "do nothing" mapper (thats what it does today).

The mappers could have an interface with name, description (used for the before mentioned list), and id. And then methods for converting the data type config and property values. The methods should simply return the new value for the config, and each property value.

So something like this:
```cs
public abstract class DataEditorConverter
{
	public string Name;
	public string Description;
	public Guid Id;

	public abstract bool IsConverter(string fromAlias, string toAlias);
	public abstract object ConvertConfig(object fromConfig);
	public abstract object ConvertValue(object fromValue);
}
```

Which then could be implemented like this:
```cs
public class NestedContentToBlockListConverter : DataEditorConverter
{
	public NestedContentToBlockListConverter()
	{
		Name = "Convert To Block List";
		Description = "Converts the config of this data type, and the property values using this data type to work with Block List";
		Id = new Guid("610fcb52-641d-4f43-9455-23a2ce674136");
	}

	public override bool IsConverter(string fromAlias, string toAlias)
	{
		return fromAlias == Constants.PropertyEditors.Aliases.NestedContent && toAlias == Constants.PropertyEditors.Aliases.BlockList;
	}

	public override object ConvertConfig(object fromConfig)
	{
		var newObject = new BlockListConfiguration();
		// do stuff that converts fromConfig to an object matching the config of Block List
		return newObject;
	}

	public override object ConvertValue(object fromValue)
	{
		var newObject = new Dictionary<string, string>();
		// do stuff that converts fromValue to an object matching the value of Block List
		return newObject;
	}
}
```

In the overlay, there should be a checkbox to have the user confirm, that they are aware of doing something, that might not be undoable, and then a Save and a Cancel button.

**Reply by DanDiplo** [reply]
@skttl I think that's brilliant and definitely along the lines I was thinking myself. I believe if Umbraco HQ can provide the underlying mechanism for this then I'm sure the community could write implementations of specific converters. Often it's not that difficult to transform one model to another - what is difficult is providing the UI and having something deal with the "plumbing" of dialogues, saving etc. But if HQ could provide the "plumbing", I'd be more than happy to write converters (and I'm sure other package devs would be happy to, too). People like @KevinJump have a lot of experience in this, I believe!

You could even have converters for more simple cases, like changing a rich-text to a textbox, where the converter just strips out the HTML tags. Or for upgrading older property editors that might store content as Ids and convert them to UDIs etc. There are a lot of possibilities 👍

**Reply by NikRimington** [reply]
I like this idea, but there could potentially be alot of converters that are needed, or a way of indicating that "This type cannot be converted to that type" mechanism to prevent people changing properties from things like nested content to numeric/url pickers etc.

**Reply by skttl** [reply]
@NikRimington that is where the "Reset" and "Do nothing" converters come in. The idea is to present a selection of converters and let the user pick what fits their needs - which also could end up having the user cancel changing the property editor.

So we could start out with only some (like mediapicker2 to 3, nested content to block list), and then grow the list or let people implement their own.

One problem though - uSync and Deploy needs some way of knowing what converter to run.

**Reply by marcemarc** [reply]
When does the conversion take place?

Do we envisage a thread going off and converting all the stored content for the property in the database, for all variants, and versions? and then republishing?

I guess it depends on the change...

eg MediaPicker V2 -> V3, would work with an EditorModelManager Sending Event to convert the old format to new format when opened by the editor, so no data is lost - and a PropertyValueConverter that would know if it was the V2 format or the V3 format it was working with...

... but probably not possible with complicated stuff!

**Reply by skttl** [reply]
The first one, which is why I see challenges with deploy/usync.

**Reply by patrickdemooij9** [reply]
But wouldn't deploy/usync be able to use the IsConverter function to see what converter to run?

**Reply by skttl** [reply]
@patrickdemooij9 not in my example, where multiple converters can be chosen from.

**Reply by patrickdemooij9** [reply]
@skttl, ah I understand now. The only way I could see this working for uSync or deploy is that they track what converters are used. So if you select the converter from NestedContent to BlockEditor, it'll save it (maybe key, alias, something?) in the uSync file, so that it can execute that on importing.

Or we have to offer the same option when importing the data, but then you'll have to remember what converter you chose.

**Reply by ronaldbarendse** [reply]
All property/data editors can already return a custom [IDataValueEditor](https://github.com/umbraco/Umbraco-CMS/blob/40dbfa7b8846c2835ddce08cdf690a9dc8719a94/src/Umbraco.Core/PropertyEditors/IDataValueEditor.cs) that allows changing the value when it's sent from the database to the editor and vice versa.
